File: Software/C/Examples/motors.c
/*
 *  https://www.dexterindustries.com/BrickPi/
 *  https://github.com/DexterInd/BrickPi3
 *
 *  Copyright (c) 2026 Modular Robotics Inc
 *  Released under the MIT license (http://choosealicense.com/licenses/mit/).
 *  For more information, see https://github.com/DexterInd/BrickPi3/blob/master/LICENSE.md
 *
 *  This code is an example for reading the encoders of motors connected to the BrickPi3.
 *
 *  Hardware: Connect EV3 or NXT motor(s) to any of the BrickPi3 motor ports.
 *
 *  Results:  When you run this program, you should see the encoder value for each motor.
 * By manually rotating motor A, the other motor(s) will be controlled.
 * Motor B power will be controlled, Motor C speed will be controlled,
 * and motor D position will be controlled.
 *
 *  Example compile command:
 *    g++ -o program "motors.c"
 *  Example run command:
 *    sudo ./program
 *
 */

#include "../BrickPi3.cpp" // for BrickPi3
#include <stdio.h>      // for printf
#include <unistd.h>     // for usleep
#include <signal.h>     // for catching exit signals

BrickPi3 BP;

void exit_signal_handler(int signo);

int main(){
  signal(SIGINT, exit_signal_handler); // register the exit function for Ctrl+C

  BP.detect(); // Make sure that the BrickPi3 is communicating and that the firmware is compatible with the drivers.

  // Reset the encoders
  BP.reset_motor_encoder(PORT_A + PORT_B + PORT_C + PORT_D);

  // Veriables for reading motor state bits
  uint8_t StateA, StateB, StateC, StateD;

  // Variables for reading motor powers
  int8_t PowerA, PowerB, PowerC, PowerD;

  // Variables for reading motor encoder positions
  int32_t PositionA, PositionB, PositionC, PositionD;

  // Variables for reading motor speeds (Degrees Per Second)
  int16_t DPSA, DPSB, DPSC, DPSD;

  while(true){
    // Read the motor status values
    BP.get_motor_status(PORT_A, StateA, PowerA, PositionA, DPSA);
    BP.get_motor_status(PORT_B, StateB, PowerB, PositionB, DPSB);
    BP.get_motor_status(PORT_C, StateC, PowerC, PositionC, DPSC);
    BP.get_motor_status(PORT_D, StateD, PowerD, PositionD, DPSD);

    // Use the encoder value from motor A to control motors B, C, and D
    BP.set_motor_power(PORT_B, PositionA < 100 ? PositionA > -100 ? PositionA : -100 : 100);
    BP.set_motor_dps(PORT_C, PositionA);
    BP.set_motor_position(PORT_D, PositionA);

    // Display the encoder values
    printf("State A: %d  B: %d  C: %d  D: %d  Power A: %4d  B: %4d  C: %4d  D: %4d  Encoder A: %6d  B: %6d  C: %6d  D: %6d  DPS A: %6d  B: %6d  C: %6d  D: %6d\n", StateA, StateB, StateC, StateD, PowerA, PowerB, PowerC, PowerD, PositionA, PositionB, PositionC, PositionD, DPSA, DPSB, DPSC, DPSD);

    // Delay for 20ms
    usleep(20000);
  }
}

// Signal handler that will be called when Ctrl+C is pressed to stop the program
void exit_signal_handler(int signo){
  if(signo == SIGINT){
    BP.reset_all();    // Reset everything so there are no run-away motors
    exit(-2);
  }
}

File: Software/Rust/brickpi3/examples/motors.rs
//
// This code is an example for reading the encoders of motors connected to the BrickPi3.
//
// Copyright (C) 2017 Juhana Helovuo <juhe@iki.fi>
// Released under the MIT license (http://choosealicense.com/licenses/mit/).
//
// Hardware: Connect EV3 or NXT motor(s) to any of the of the BrickPi3 motor ports.
//
// Results:  When you run this program, you should see the encoder value for
// each motor. By manually rotating motor A, the other motor(s) will be
// controlled. Motor B power will be controlled, Motor C speed will be
// controlled, and motor D position will be controlled.
//
// Compilation command, in the crate top directory (brickpi3)
//   cargo build --example motors

extern crate brickpi3;
extern crate ctrlc;
extern crate strum; // strum is for iterating over enums
// #[macro_use] extern crate strum_macros;

use std::{thread, time};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use strum::IntoEnumIterator;
use brickpi3::*;

fn main() {
	match brickpi3::BrickPi3::open("/dev/spidev0.1") {
		Err(why) =>  println!("Opening BrickPi3 SPI device failed: {:?}", why) ,
		Ok(mut bp) => {
			match bp.detect() {
				Err(why) => println!("BrickPi3 detection failed: {:?}", why) ,
				Ok(_ident) => {
					match run_motors(&mut bp) {
						Err(why) => { println!("BrickPi error: {:?}", why);
									  bp.reset_all()
									  	.unwrap_or_else(|e| println!("{:?}",e))	
									},
						Ok(()) => ()
					}
				}
			}
		}
	}
}

fn run_motors(bp: &mut BrickPi3) -> Result<(),SPIError> {
	// Set signal handler for Ctrl-C
	let running = Arc::new(AtomicBool::new(true));
    let r = running.clone();
    ctrlc::set_handler(move || {
        r.store(false, Ordering::SeqCst);
    }).expect("Error setting Ctrl-C handler");

    // reset all motor encoders
    for mp in MotorPort::iter() {
    	bp.reset_motor_encoder(mp)
    		.unwrap_or_else(|e| {println!("{:?}",e); 0} );
    }

    // loop until interrupted
    while running.load(Ordering::SeqCst) {
    	let enc_a = bp.read_motor_encoder(MotorPort::PortA) ?;
    	let enc_b = bp.read_motor_encoder(MotorPort::PortB) ?;
    	let enc_c = bp.read_motor_encoder(MotorPort::PortC) ?;
    	let enc_d = bp.read_motor_encoder(MotorPort::PortD) ?;

    	bp.set_motor_power(	MotorPort::PortB
    						, if enc_a < 100 
    							{if enc_a > -100 {enc_a as i8} else {-100}} 
    						  else {100} 
    						) ?;
    	bp.set_motor_dps( MotorPort::PortC, enc_a as i16) ? ;
    	bp.set_motor_position(MotorPort::PortD , enc_a) ? ;

    	println!("Encoders A: {:.2 } B: {:.2 } C: {:.2 } D: {:.2 }", enc_a, enc_b, enc_c, enc_d);

    	// wait 200ms before next iteration
    	thread::sleep(time::Duration::from_millis(200));

    }
    // ensure no motors are left running
    bp.reset_all()
}

	
    
    